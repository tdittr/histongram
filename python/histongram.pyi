from typing import Iterator


class Histogram:
    def __init__(self): ...

    def add(self, key: str): ...

    def add_many(self, iter: Iterator[str]): ...

    def __getitem__(self, key: str) -> int: ...
