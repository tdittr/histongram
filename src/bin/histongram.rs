use std::fs::read_to_string;
use std::path::PathBuf;

use clap::Parser;
use color_eyre::Result;
use compact_str::CompactString;

use histongram::{Histogram, Ngrams};

#[derive(Parser, Debug, Clone)]
struct Args {
    file: PathBuf,

    #[clap(long)]
    print: bool,
}

fn main() -> Result<()> {
    color_eyre::install()?;
    let args = Args::parse();

    let data = read_to_string(args.file)?;

    let ngrams = Ngrams::new(1..=5).count(data.split_whitespace());

    if args.print {
        println!("{ngrams:?}");
    }

    Ok(())
}
