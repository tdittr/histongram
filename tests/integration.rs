use std::cmp::{min, Reverse};

use insta::assert_ron_snapshot;

use histongram::{DefaultHashBuilder, Histogram};

const APACHE: &str = include_str!("../LICENSE-APACHE");
const MIT: &str = include_str!("../LICENSE-MIT");

#[test]
fn license_words() {
    let mut a: Histogram<&'static str> = APACHE.split_whitespace().collect();
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m: Histogram<_> = MIT.split_whitespace().collect();
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[test]
fn license_chars() {
    let mut a = Histogram::<_, DefaultHashBuilder>::from_owned_iter(APACHE.chars());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_owned_iter(MIT.chars());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[cfg(feature = "serde")]
#[test]
fn serde() {
    let mut settings = insta::Settings::clone_current();
    settings.set_sort_maps(true);
    settings.bind(|| {
        let mut a = Histogram::<_, DefaultHashBuilder>::from_owned_iter(APACHE.chars());
        assert_ron_snapshot!(a);

        let m = Histogram::<_, DefaultHashBuilder>::from_owned_iter(MIT.chars());
        assert_ron_snapshot!(m);

        a.append(m);
        assert_ron_snapshot!(a);
    });
}

fn sort_also_by_key<K: Ord + Copy>(counts: &mut [(K, usize)]) {
    // This table should be sorted by count. But items with the same number of occurrences are
    // in arbitrary order...
    let mut high = usize::MAX;
    for &(_, cnt) in counts.iter() {
        assert!(cnt <= high);
        high = min(high, cnt);
    }

    // Now sort also by key so the snapshot stays consistent
    counts.sort_unstable_by_key(|(key, cnt)| (Reverse(*cnt), *key));
}
