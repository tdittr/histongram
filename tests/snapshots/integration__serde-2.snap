---
source: tests/integration.rs
expression: m
---
{
  '
': 16,
  ' ': 147,
  '"': 4,
  '(': 1,
  ')': 1,
  ',': 22,
  '.': 3,
  '/': 1,
  ':': 1,
  'A': 28,
  'B': 5,
  'C': 10,
  'D': 10,
  'E': 35,
  'F': 12,
  'G': 6,
  'H': 18,
  'I': 35,
  'K': 1,
  'L': 14,
  'M': 7,
  'N': 30,
  'O': 36,
  'P': 8,
  'R': 34,
  'S': 25,
  'T': 39,
  'U': 8,
  'V': 2,
  'W': 9,
  'X': 1,
  'Y': 6,
  'a': 25,
  'b': 9,
  'c': 16,
  'd': 16,
  'e': 46,
  'f': 15,
  'g': 8,
  'h': 20,
  'i': 43,
  'j': 1,
  'l': 16,
  'm': 8,
  'n': 30,
  'o': 49,
  'p': 11,
  'r': 25,
  's': 33,
  't': 47,
  'u': 12,
  'v': 1,
  'w': 10,
  'y': 6,
}
